src/main/java/com/iota/iri/hash/PearlDiver.java

```diff
@@ -52,135 +52,46 @@ public synchronized boolean search(final int[] transactionTrits, final int minWe
-            midCurlStateLow[162 + 2] = 0b0111111111111111111000000000111111111111111111000000000111111111L;
-            midCurlStateHigh[162 + 2] = 0b1111111111000000000111111111111111111000000000111111111111111111L;
-            midCurlStateLow[162 + 3] = 0b1111111111000000000000000000000000000111111111111111111111111111L;
-            midCurlStateHigh[162 + 3] = 0b0000000000111111111111111111111111111111111111111111111111111111L;
-
-        }
-
+        para(transactionTrits, midCurlStateLow, midCurlStateHigh);
+        
         int numberOfThreads = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
 
         Thread[] workers = new Thread[numberOfThreads];
 
         while (numberOfThreads-- > 0) {
-
             final int threadIndex = numberOfThreads;
+            
             Thread worker = (new Thread(() -> {
-
                 final long[] midCurlStateCopyLow = new long[CURL_STATE_LENGTH], midCurlStateCopyHigh = new long[CURL_STATE_LENGTH];
+                
                 System.arraycopy(midCurlStateLow, 0, midCurlStateCopyLow, 0, CURL_STATE_LENGTH);
                 System.arraycopy(midCurlStateHigh, 0, midCurlStateCopyHigh, 0, CURL_STATE_LENGTH);
+                
                 for (int i = threadIndex; i-- > 0; ) {
                     increment(midCurlStateCopyLow, midCurlStateCopyHigh, 162 + CURL_HASH_LENGTH / 9,
                             162 + (CURL_HASH_LENGTH / 9) * 2);
 
                 }
 
                 final long[] curlStateLow = new long[CURL_STATE_LENGTH], curlStateHigh = new long[CURL_STATE_LENGTH];
-                final long[] curlScratchpadLow = new long[CURL_STATE_LENGTH], curlScratchpadHigh = new long[CURL_STATE_LENGTH];
                 long mask, outMask = 1;
-                while (state == RUNNING) {
-
-                    increment(midCurlStateCopyLow, midCurlStateCopyHigh, 162 + (CURL_HASH_LENGTH / 9) * 2,
-                            CURL_HASH_LENGTH);
-
-                    System.arraycopy(midCurlStateCopyLow, 0, curlStateLow, 0, CURL_STATE_LENGTH);
-                    System.arraycopy(midCurlStateCopyHigh, 0, curlStateHigh, 0, CURL_STATE_LENGTH);
-                    transform(curlStateLow, curlStateHigh, curlScratchpadLow, curlScratchpadHigh);
-
-                    mask = HIGH_BITS;
-                    for (int i = minWeightMagnitude; i-- > 0; ) {
-                        mask &= ~(curlStateLow[CURL_HASH_LENGTH - 1 - i] ^ curlStateHigh[
-                            CURL_HASH_LENGTH - 1 - i]);
-                        if (mask == 0) {
-                            break;
-                        }
-                    }
-                    if (mask == 0) {
-                        continue;
-                    }
-
-                    synchronized (syncObj) {
-                        if (state == RUNNING) {
+                
+                if ((mask = loop(midCurlStateCopyLow, midCurlStateCopyHigh, minWeightMagnitude)) != 0) {
```

Comment:
Check quality review: https://app.codacy.com/app/alon-e/iri/pullRequest?prid=1361219

Reply:
I’ve noticed that, and also thinking how to improve NPath complexity